### PR TITLE
support unbanlanced pp for vllm.

### PR DESCRIPTION
**Pull request by charles9304**

### CODE REVIEW

File: chatlearn/utils/vllm_utils.py [hunk 1]
```diff
@@ -821,8 +826,16 @@ def convert_llama_state_dict_from_megatron_to_vllm(args, hf_config, qwen_version
 
     # Transformer Layers
     print("Converting transformer layers")
-    num_layers = hf_config.num_hidden_layers // pp_size
-
+    if CURRENT_VLLM_VERSION == VLLMVersion.v_0_6_3:
+        start_layer_idx, _ = get_pp_indices(
```
Comment (adoda) [inline chatlearn/utils/vllm_utils.py:830]: we needed build pp_indices from  the mapping by training to inference，and then set envs.VLLM_PP_LAYER_PARTITION for vllm.

Comment (charles9304) [reply chatlearn/utils/vllm_utils.py:830]: Done. We need to set envs.VLLM_PP_LAYER_PARTITION before model setup, With config num_layers and pipeline_model_parallel_size, we infer num of layers for each stage when allow_padding_num_layers set True.

File: chatlearn/models/vllm_module.py [hunk 2]
```diff
@@ -252,6 +257,7 @@ def init(self):
         initialize_vllm(extra_args_provider=self.add_extra_args,
                         ignore_unknown_args=True,
                         args_dict=self.model_args)
+        self.parallel_config.rank = torch.distributed.get_rank()
```
Comment (adoda) [inline chatlearn/models/vllm_module.py:260]: where used?

Comment (charles9304) [reply chatlearn/models/vllm_module.py:260]: When profiling gpu blocks, model_runner in vLLM worker uses to get pp layer partition. 

ref: 

1. https://github.com/vllm-project/vllm/blob/v0.6.3.post1/vllm/worker/model_runner.py#L1276
2. https://github.com/vllm-project/vllm/blob/v0.6.3.post1/vllm/config.py#L549

File: chatlearn/models/vllm_module.py [hunk 3]
```diff
@@ -832,6 +838,50 @@ def onload_weights(self):
...
+
+        # Each stage gets a contiguous set of layers.
+        if self.model_args.get("pipeline_layers", None) is not None:
+            rank_sizes = self.model_args.get("pipeline_layers", None)
+            assert isinstance(rank_sizes, list) and all(isinstance(ele, int) for ele in rank_sizes), \
+                f"pipeline_layers expected to be list, and num layer of each stage to be integer, while {rank_sizes}."
+        else:
+            rank_sizes = [num_layers_per_stage_with_padding] * pipeline_world_size
+            num_padding = num_layers - num_layers_without_padding
+            for _index in range(-1, num_padding - 1):
```
Comment (adoda) [inline chatlearn/models/vllm_module.py:872]: We need to redesign the splitting logic when num_padding != 2, we can create a new CR to handle it.

**Review by adoda — APPROVED** [review]

File: chatlearn/models/vllm_module.py [hunk 4]
```diff
@@ -832,6 +838,50 @@ def onload_weights(self):
...
+            num_layers // pipeline_world_size)
+
+        # Each stage gets a contiguous set of layers.
+        if self.model_args.get("pipeline_layers", None) is not None:
+            rank_sizes = self.model_args.get("pipeline_layers", None)
+            assert isinstance(rank_sizes, list) and all(isinstance(ele, int) for ele in rank_sizes), \
+                f"pipeline_layers expected to be list, and num layer of each stage to be integer, while {rank_sizes}."
+        else:
+            rank_sizes = [num_layers_per_stage_with_padding] * pipeline_world_size
+            num_padding = num_layers - num_layers_without_padding
```
Comment (adoda) [inline chatlearn/models/vllm_module.py:871]: we can add a check here, and prompt the user that they can configure pipeline_layers

Comment (charles9304) [reply chatlearn/models/vllm_module.py:870]: Done. but may not tolerant enough, should support to inference with train_pp later.

File: chatlearn/models/vllm_module.py [hunk 5]
```diff
@@ -832,6 +838,50 @@ def onload_weights(self):
...
+            for _index in range(-1, num_padding - 1):
+                rank_sizes[_index] -= 1
+        assert len(rank_sizes) == pipeline_world_size
+
+        # set evn variable VLLM_PP_LAYER_PARTITION
```
[the remark this answers is not in the document]
Comment (charles9304) [reply chatlearn/models/vllm_module.py]: done

File: chatlearn/utils/vllm_utils.py [hunk 6]
```diff
@@ -1292,14 +1315,16 @@ def convert_qwen_state_dict_from_megatron_to_vllm(args, hf_config, qwen_version=
...
                 for params in params_list:
                     params = params.view((moe_num_experts, -1, hf_config.hidden_size)).contiguous()
                     params = params.reshape((moe_num_experts // tp_size * 2, -1, hf_config.hidden_size))
                     params = params.chunk(tp_size, dim=1)[tp_rank]
-                    params = params.reshape(params.shape[0] // tp_size * 2, -1, hf_config.hidden_size)
+                    params = params.reshape(params.shape[0] // 2, -1, hf_config.hidden_size)
                     params_right, params_left = params.chunk(2, dim=1)
-                    params = torch.cat([params_left, params_right], dim=1)
+                    params = torch.cat([params_left, params_right], dim=1).view(
+                        moe_num_experts // tp_size, -1, hf_config.hidden_size).contiguous()
```
[the remark this answers is not in the document]
Comment (charles9304) [reply chatlearn/utils/vllm_utils.py]: Revert already. Thanks

**Review by haolin-nju — APPROVED** [review]
LGTM

**Review by haolin-nju — APPROVED** [review]
LGTM. I have a small suggestion: perhaps we should add daily regression tests on unbalanced pp for vllm.

**Review by adoda — APPROVED** [review]

**Pull request merged**